.github/workflows/ci-test.yml

```diff
@@ -63,15 +102,44 @@ jobs:
         key: test-${{ runner.os }}-clang
         max-size: 1000M
 
-    - name: Build (Ubuntu Linux, Clang, Sanitizers)
+    - name: Build (Ubuntu 20.04, Clang, Sanitizers)
+      run: |
+        sudo cp -rf . /p4c/
+        (cd /p4c/ && sudo -E tools/ci-build.sh)
+        sudo cp -rf /p4c/.ccache .
+
+    - name: Run tests (Ubuntu 20.04)
+      run: sudo -E ctest --output-on-failure --schedule-random
+      working-directory: /p4c/build
+
+  # Build with gcc and test p4c on Ubuntu 18.04.
+  test-ubuntu18:
+    strategy:
+      fail-fast: false
+    runs-on: ubuntu-latest
+    env:
+      CTEST_PARALLEL_LEVEL: 4
+      ENABLE_UNIFIED_COMPILATION: ON
+    steps:
+    - uses: actions/checkout@v2
+      with:
+        submodules: recursive
+        fetch-depth: 0
+
+    - name: ccache
+      uses: hendrikmuhs/ccache-action@v1
+      with:
+        key: test-ubuntu-18.04-gcc
+        max-size: 1000M
+
+    - name: Build (Ubuntu 18.04, GCC)
       run: |
-        docker build -t p4c --build-arg IMAGE_TYPE=test --build-arg ENABLE_UNIFIED_COMPILATION=ON --build-arg COMPILE_WITH_CLANG=ON \
-        --build-arg BUILD_AUTO_VAR_INIT_PATTERN=ON --build-arg ENABLE_SANITIZERS=ON .
+        docker build -t p4c --build-arg BASE_IMAGE=ubuntu:18.04 --build-arg IMAGE_TYPE=test --build-arg ENABLE_UNIFIED_COMPILATION=ON .
```

Comment:
This also needs to be updated to use Ubuntu 18.04 directly. Also other images such as the tools test should not use docker anymore then.

Reply:
There is no Ubuntu 18.04 image for github actions. So nope, not possible.

Reply:
>> Also other images such as the tools test should not use docker anymore then.

Sorry but for now I am not gonna do everything. We can move it incrementally.

Reply:
But we do not modify the base image right now so this will test on the Ubuntu 20.04 docker image as far as I can see.

Reply:
> > > Also other images such as the tools test should not use docker anymore then.
> 
> Sorry but for now I am not gonna do everything. We can move it incrementally.

The issue is that the new ci-build configuration seems to break the ccache that has been set up for the docker build process. As far as I can tell, the builds are taking longer now. We can check this by completing one build and rerunning to see whether the build is cached.

Reply:
Build on Ubuntu 22 - 6m 14s.

Seems pretty OK to me.

Reply:
Comparing it with for example https://github.com/p4lang/p4c/pull/3778 I see nothing problematic.

Reply:
The problem is the other tests, `static-build-test-p4c`,  `test-tools`,  `ptf-ebpf`, `validate` etc.

Reply:
>> the builds are taking longer now. 

Sure?

test-p4c / test-ubuntu20 (OFF) (pull_request) Successful in 5m
test-p4c / test-ubuntu20 (ON) (pull_request) Successful in 37m

-------------------------------------------

https://github.com/p4lang/p4c/pull/3778

test-p4c / test-linux (OFF) (pull_request) Successful in 5m
test-p4c / test-linux (ON) (pull_request) Successful in 69m

Reply:
ptf-ebpf / ptf-linux (pull_request) Successful in 62m 

https://github.com/p4lang/p4c/pull/3778
 ptf-ebpf / ptf-linux (pull_request) Successful in 18m

Reply:
Yes, but this PR is breaking the ccache setup for all other tests. 

The reason `validate` is faster is that the tests are not running for some reason, that seems like an unrelated bug. 

In general, `test-tools` will now be the slowest test because the build alone takes 55 minutes.

Reply:
>> The problem is the other tests, static-build-test-p4c, test-tools, ptf-ebpf, validate etc.

I can convert them but then we would stop testing with Dockerfile so it can slowly bitrot.

Reply:
> Yes, but this PR is breaking the ccache setup for all other tests.

How? Could you explain it?

Docker builds uses export_ccache.sh which runs
docker cp $(docker create --rm p4c):/p4c/.ccache .
, non-docker runs 
sudo cp -rf /p4c/.ccache .

I pushed empty commit to see if regressions are real.

Reply:
Could be [this](https://github.com/p4lang/p4c/pull/3770/files#diff-76f06548e16be4f20ba423b3768e29bfca82a8ecd274407c02b37b3b5816f0c8L57) change.

Reply:
> > > The problem is the other tests, static-build-test-p4c, test-tools, ptf-ebpf, validate etc.
> 
> I can convert them but then we would stop testing with Dockerfile so it can slowly bitrot.

We can either convert them or have a workaround. As Radostin [mentioned](https://github.com/p4lang/p4c/pull/3770#discussion_r1045273096) we need to create one docker build either way.

Reply:
>> In general, test-tools will now be the slowest test because the build alone takes 55 minutes.

ccache is working fine, rebuild time: 7m 25s

https://github.com/p4lang/p4c/actions/runs/3697344554/jobs/6262220551